### PR TITLE
Persist job runs past max_prowjob_age by storing them in the GH status description

### PR DESCRIPTION
**Pull request by alvaroaleman**
Currently, Tide always has to retest if a test result is older than
max_prowjob_age because it can not determine anymore what baserevision
the test run ran against.

While it is possible to optimize Sinker to be able to keep presubmits
longer, the bigger Prow instances like the Kubernetes or Openshift one
will never be able to keep ProwJobs for more than a couple of days.
This is not sufficient for for example backports, where PRs regularly
wait for a week or more until they can be merged.

This PR makes us (ab-)use the GitHub status context description to store
the baseSHA and makes Tide's accumulate construct virtual ProwJob
objects out of these. This allows us to store past presubmit results
pretty much indefinitely.

Apart from the included unittests I also manually deployed this and verified that Tide merges without creating a new Prowjob when the baseSHA is current and continues to create the ProwJob if not.

The status now looks like this on GitHub:
gh_basesha

This approach was discussed in Slack: https://kubernetes.slack.com/archives/CDECRSC5U/p1616509899019900
Supersedes and thereby closes https://github.com/kubernetes/test-infra/pull/21453

I recommend disabling whitespace changes to reduce the diff. Most of the change in this PR is tests, because the PR object tide uses has very deeply nested substructs.

/assign @chaodaiG @cjwagner @stevekuznetsov

### CODE REVIEW

File: prow/tide/tide.go
Comment (stevekuznetsov) [inline prow/tide/tide.go:958]: This is a bit confusing - instead of only creating ProwJobs that pass against the current HEAD (I assume we append this to the filtered list?) why don't we just always create the synthetic ProwJob and have it specify the base ref it tested against, then have the filtering etc happen as normal after?

Comment (alvaroaleman) [reply prow/tide/tide.go:958]: Because the list this gets appended to is already filtered by baseSHA

File: prow/config/config.go
Comment (chaodaiG) [inline prow/config/config.go:2558]: This assumes `Basesha:<SHA>` always the last element of context description, is there test guarding this?

File: prow/tide/tide.go
Comment (chaodaiG) [inline prow/tide/tide.go:978]: can we use `log` instead of `logrus` for more info?

File: prow/tide/tide.go
Comment (chaodaiG) [inline prow/tide/tide.go:980]: I think appending means the priority of status context is higher than prowjobs, so if status context says pass and prowjobs says fail, tide will still merge the PR. Is there any potential risk here? For example outdated status context?

File: prow/config/config.go
Comment (alvaroaleman) [reply prow/config/config.go:2558]: There is `TestContextDescriptionWithBaseShaRoundTripping` and we will only return it if the second element has the correct length (40 characters). The worst case for this is that this functionality breaks but not that we incorrectly merge something, because Tide checks if the basesha is correct

File: prow/tide/tide.go
Comment (alvaroaleman) [reply prow/tide/tide.go]: good catch, fixed

File: prow/tide/tide.go
Comment (alvaroaleman) [reply prow/tide/tide.go:980]: The status context are retrieved for a sha, so they can not be outdated (except for the basesha which is why we are doing the dance here). Tide always tries to find the best result and will use that (i.E. if you have one failure, one success, the job is considered a success) so the order shouldn't be relevant: https://github.com/kubernetes/test-infra/blob/eeb7afb52810b459ddc23b33890e69cbf8efb61f/prow/tide/tide.go#L983

**Comment by alvaroaleman** [comment]
Urg #21507
/retest

File: prow/tide/tide.go
Comment (chaodaiG) [reply prow/tide/tide.go:980]: I see, this makes sense. Since they are the same sha and baseSha, the order shouldn't matter. Even in worst case, say a flaky job passed then failed, tide will still respect the passed context if the failed job hasn't been reported by crier yet, which is totally fine with me, and I would even consider it a bonus. So SGTM

**Review by chaodaiG — APPROVED** [review]
/hold

File: prow/tide/tide.go
Comment (stevekuznetsov) [reply prow/tide/tide.go:958]: Hm - I guess this entire change would be less dangerous etc if we always created fake ProwJobs with whatever baseSHA and status we see in GitHub, then we'd only have one filtering code for the list and not potentially introduce two areas that, if bugged, have large impacts on tide.

File: prow/tide/tide.go
Comment (stevekuznetsov) [reply prow/tide/tide.go:980]: Yeah see like *this* conversation also does not have to happen if we generate all contexts with correct timestamps etc and do filtering after.

**Review by stevekuznetsov — APPROVED** [review]
The code as written looks correct. While I still think there's a more defensive way to code this, if we're happy with this impl please feel free to cancel the hold and merge.

/hold

File: prow/tide/tide.go
Comment (alvaroaleman) [reply prow/tide/tide.go]: Updated it to create ProwJobs instead. It will still only create prowjobs for the matching baseSHA and successes to not overload the apiserver, but it will take the created baseSHA from the github context.

Comment (cjwagner) [reply prow/tide/tide.go]: Why do the ProwJobs need to be added to the API server? That seems like unnecessary load on the API server given that Tide is the only thing that will consume these. There are also other catches like avoiding creating the same PJs every sync loop and ensuring other Prow components handle the fake PJs appropriately (e.g. Deck should ignore them, sinker would need to handle them differently).
I interpreted Steve's request to mean create ProwJob structs so that we can filter them the same way we filter real PJs, not create real ProwJob CRs on the API server.

Comment (alvaroaleman) [reply prow/tide/tide.go:958]: > Why do the ProwJobs need to be added to the API server?

Because the filtering is done via [an index over the cache](https://github.com/kubernetes/test-infra/blob/eeb7afb52810b459ddc23b33890e69cbf8efb61f/prow/tide/tide.go#L1716) in order to be efficient and I don't think trying to trick the cache into thinking there was data that isn't on the server is a great idea.

> There are also other catches like avoiding creating the same PJs every sync loop and ensuring other Prow components handle the fake PJs appropriately 

True, I've reverted the pull to the original version

> I interpreted Steve's request to mean create ProwJob structs so that we can filter them the same way we filter real PJs, not create real ProwJob CRs on the API server.

It already did create prowjob structs in the first version and as mentioned above, re-using the filtering isn't really possible.

**Comment by fejta** [comment]
LGTM
/uncc
Will defer to Cole (and Chao) for this PR

**Comment by stevekuznetsov** [comment]
/lgtm
/hold cancel

The last commit reverted back to the changes that had been LGTMed in the past, this seems good to go.

**Pull request merged**
